lib/pure/asyncdispatch.nim

```diff
@@ -1851,6 +1907,105 @@ proc send*(socket: AsyncFD, data: string,
 
   return retFuture
 
+proc deInitVirtualEventDispatcher(p: PDispatcher) =
+  assert p.vd.virtualHandles.len == 0, "Cannot de-Init Virtual Event Dispacter. There are still Pending Events."
+  p.vd.virtualMuxHandle.unregister()
+  p.vd.virtualMuxHandle.close()
+  p.vd.virtualMuxHandle = nil
+  p.vd.nextVirtualHandle = 0.VirtualFD
+
+proc unregister*(ev: VirtualAsyncEvent) =
+  ## Unregisters event ``ev``.
+  doAssert(ev.vFD != InvalidVirtualFD, "Event is not registered in the queue!")
+
+  let oldP = ev.p
+  withLockIfThreads ev.eventLock:
+    ev.p.vd.virtualHandles.del ev.vFD
+    ev.p = nil
+    ev.cb = nil
+    ev.vFD = InvalidVirtualFD
+
+  if oldP.vd.virtualHandles.len == 0:
+    # lazy de-init the physical event with the Dispatcher
+    # The main reason we need this is to make
+    # hasPendingOperations still work
+    # without modifying the ioselector code.
+    deInitVirtualEventDispatcher(oldP)
+
+proc initVirtualEventDispatcher(p: PDispatcher) =
+  p.vd.virtualMuxHandle = newAsyncEvent()
+
+  proc virtualMuxCB(fd: AsyncFD): bool {.closure,gcsafe.} =
+    # find the virtual events that triggered the physical event and
+    # add them to the callback list.
+    # Not very efficient, but requires the least coordination between threads.
+    
+    proc vcbFactory(vfd: VirtualFD) : proc() {.closure, gcsafe.} =
+      let ev = p.vd.virtualHandles[vfd]
+      result = proc() {.closure,gcSafe.} =
+        if ev.cb(vfd.AsyncFD):
+          # the convention is that if the callback returns true, we unregister the event.
+          ev.unregister
+
+    for vfd, ev in p.vd.virtualHandles:
+      withLockIfThreads ev.eventLock:
+        if ev.triggered:
+          ev.triggered = false
+          p.callbacks.addLast(vcbFactory(vfd))
```

Comment:
I really want `initVirtualEventDispatcher` to take a `VirtualEventDispatcher` object, and not a `PDispatcher` object.
The **main** reason I need the `PDispatcher` here is because I can think of no other way to add a virtual callback into Dispatcher callback queue.

I *could* have `VirtualEventDispatcher` maintain a separate dispatch queue. But that could cause all kinds of problems, because callback order matters. As seen here for example: #7197

It is just much safer to insert the callback into the parent `PDispatcher` queue.

Reply:
Can't you use getGlobalDispatcher?

Reply:
No. Two reasons:
1.) It's not thread safe.
2.) Related to 1, just like physical `AsyncEvents`, `VirtualAsyncEvents` are not assigned a dispatcher by default. They can technically be assigned to different dispatchers.
The API allows for this, so it has to be accounted for.

Reply:
To clarify this further:

It's ok to use getGlobalDispatcher() on creation. see:
https://github.com/nim-lang/Nim/blob/264bfeebff4ddcee4d224c9b237f3f607e5ebcea/lib/pure/asyncdispatch.nim#L1983

**But** the trigger event may be called from a different thread with a completely different dispatcher!

In that case, we need a way to trace back to the owning `PDispatcher` instance.

Reply:
I changed my mind. I thought about this more, and `getGlobalDispatcher` is thread safe (it returns a thread local global.)

I changed the code to use `getGlobalDispatcher` and removed the closure requirement.